api/bucket.go

```diff
@@ -1,19 +1,209 @@
+
+	ungzReader, err := gzip.NewReader(fileReader)
+	if err != nil {
+		ctx.Status(http.StatusBadRequest)
+		return
+	}
+	defer ungzReader.Close()
+
+	// Report 5xx errors for extraction process
+	defer func() {
+		if err != nil {
+			log.Error(err, "bucket archive expand action failed")
+			ctx.JSON(http.StatusInternalServerError, gin.H{
+				"error": err,
+			})
+			return
+		}
+	}()
+
+	// Clean and prepare destination directory
+	err = os.RemoveAll(dir)
+	if err != nil {
+		return
+	}
+	if err = os.MkdirAll(dir, 0777); err != nil {
+		return
+	}
+
+	// Extract the tar archive
+	untarReader := tar.NewReader(ungzReader)
+	for {
+		hdr, err := untarReader.Next()
+		if err != nil {
+			if err == io.EOF {
+				break
+			} else {
+				return
+			}
+		}
+
+		switch hdr.Typeflag {
+		case tar.TypeDir:
+			if err := os.Mkdir(path.Join(dir, hdr.Name), 0777); err != nil {
+				return
+			}
+		case tar.TypeReg:
+			var file *os.File
+			if file, err = os.Create(path.Join(dir, hdr.Name)); err != nil {
+				return
+			}
+			if _, err = io.Copy(file, untarReader); err != nil {
+				return
+			}
+			file.Close()
+		default:
+			// types that are not files or dirs are skipped
+		}
+	}
+
+	ctx.Status(http.StatusAccepted)
+}
+
+func (h *BucketHandler) getDirArchive(ctx *gin.Context, dir string) {
+	dirInfo, err := os.Stat(dir)
+	if os.IsNotExist(err) {
+		ctx.JSON(http.StatusNotFound, gin.H{
+			"error": "Provided directory path doesn't exist.",
+		})
+		return
+	}
+	if !dirInfo.IsDir() {
+		ctx.JSON(http.StatusBadRequest, gin.H{
+			"error": "Provided path is not a directory.",
+		})
+		return
+	}
+
+	var tarOutput bytes.Buffer
+	tarWriter := tar.NewWriter(&tarOutput)
+	err = filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		hdr, err := tar.FileInfoHeader(info, path)
+		if err != nil {
+			return err
+		}
+
+		// Scope path in archive to the given dir
+		hdr.Name = strings.Replace(path, dir, "", 1)
+		if hdr.Name == "" {
+			return nil
+		}
+
+		switch hdr.Typeflag {
+		case tar.TypeDir:
+			// Add directory header to the archive (no content)
+			if err := tarWriter.WriteHeader(hdr); err != nil {
+				return err
+			}
+		case tar.TypeReg:
+			// Add file with its content to the archive
+			if err := tarWriter.WriteHeader(hdr); err != nil {
+				return err
+			}
+			file, _ := os.Open(path)
+
+			if _, err = io.Copy(tarWriter, file); err != nil {
+				return err
+			}
+		default:
+			// Other file types like block/character device or TypeSymlink are skipped.
+			// Complete list of types: https://pkg.go.dev/archive/tar#pkg-constants
+		}
+
+		return nil
+	})
+
+	// Report 5xx errors in archive creation steps
+	defer func() {
+		if err != nil {
+			log.Error(err, "bucket archive get action failed")
+			ctx.JSON(http.StatusInternalServerError, gin.H{
+				"error": err,
+			})
+			return
+		}
+	}()
+
+	if err != nil {
+		return
+	}
+
+	if err := tarWriter.Close(); err != nil {
```

Comment:
Should this be in a defer instead?

Reply:
The Close() method writes footer to the tar archive (it "closes" the archive content, not the stream), it needs to be executed before ```io.Copy(gzWriter, fromTar)``` on line 202.